python/cudf/cudf/core/udf/function.cu

```diff
@@ -0,0 +1,555 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cfloat>
+#include <cstdint>
+
+using size_type = int;
+
+// double atomicAdd
+__device__ __forceinline__ double atomicAdd(double* address, double val)
+{
+  unsigned long long int* address_as_ull = (unsigned long long int*)address;
+  unsigned long long int old             = *address_as_ull, assumed;
+
+  do {
+    assumed = old;
+    old =
+      atomicCAS(address_as_ull, assumed, __double_as_longlong(val + __longlong_as_double(assumed)));
+
+  } while (assumed != old);
+
+  return __longlong_as_double(old);
+}
+
+// int64_t atomicAdd
+__device__ __forceinline__ int64_t atomicAdd(int64_t* address, int64_t val)
+{
+  return atomicAdd((unsigned long long*)address, (unsigned long long)val);
+}
+
+// double atomicMax
+__device__ __forceinline__ double atomicMax(double* address, double val)
+{
+  unsigned long long old = __double_as_longlong(*address);
+  while (val > __longlong_as_double(old)) {
+    unsigned long long assumed = old;
+    if ((old = atomicCAS((unsigned long long*)address, assumed, __double_as_longlong(val))) ==
+        assumed)
+      break;
+  }
+  return __longlong_as_double(old);
+}
+
+// int64_t atomicMax
+__device__ __forceinline__ int64_t atomicMax(int64_t* address, int64_t val)
+{
+  return atomicMax((long long*)address, (long long)val);
+}
+
+// double atomicMin
+__device__ __forceinline__ double atomicMin(double* address, double val)
+{
+  unsigned long long old = __double_as_longlong(*address);
+  while (val < __longlong_as_double(old)) {
+    unsigned long long assumed = old;
+    if ((old = atomicCAS((unsigned long long*)address, assumed, __double_as_longlong(val))) ==
+        assumed)
+      break;
+  }
+  return __longlong_as_double(old);
+}
+
+// int64_t atomicMin
+__device__ __forceinline__ int64_t atomicMin(int64_t* address, int64_t val)
+{
+  return atomicMin((long long*)address, (long long)val);
+}
+
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ T device_sum(T const* data, int const items_per_thread, size_type size) {
+  __shared__ T sum;
+  int tid = threadIdx.x;
+  int tb_size = blockDim.x;
+  T local_sum = 0;
+
+  if (tid == 0) sum = 0;
+
+  __syncthreads();
+
+// Calculate local sum for each thread
+#pragma unroll
+  for (size_type item = 0; item < items_per_thread; item++) {
+    if (tid + (item * tb_size) < size) {
+      T load = data[tid + item * tb_size];
+      local_sum += load;
+    }
+  }
+
+  atomicAdd(&sum, local_sum);
+
+  __syncthreads();
+
+  return sum;
+}
+
+// Use a C++ templated __device__ function to implement the body of the algorithm.
+template <typename T>
+__device__ T device_var(T const* data, int const items_per_thread, size_type size) {
+
+  int tid = threadIdx.x;
+  int tb_size = blockDim.x;
+
+  double local_var            = 0;
```

Comment:
Maybe using 
```suggestion
  T local_var            = 0;
```
`T` instead of `double` may save some time when dealing with single-precision `var`.

Reply:
So for this issue, if I put T instead of double, and if the data type is integer, this could cause an error since variance has to be a floating point number